tempesta_fw/http.c

```diff
@@ -382,6 +370,396 @@ tfw_http_send_504(TfwHttpReq *req)
+			continue;
+		}
+		srv_conn->msg_sent = (TfwMsg *)req;
+		/* Stop sending if the request is non-idempotent. */
+		if (tfw_http_req_is_nonidempotent(req)) {
+			TFW_DBG2("%s: Break on non-idempotent: req=[%p]\n",
+				 __func__, req);
+			break;
+		}
+		/* See if a non-idempotent request has become idempotent. */
+		tfw_http_req_nonidemp_delist(srv_conn, req);
+	}
+}
+
+/*
+ * Forward stalled requests in server connection @srv_conn.
+ *
+ * This function expects that the queue in the server connection
+ * is locked. The queue in unlocked inside the function which is
+ * very non-traditional. Please use with caution.
+ */
+static void
+tfw_http_req_fwd_stalled(TfwConnection *srv_conn)
+{
+	LIST_HEAD(equeue);
+
+	TFW_DBG2("%s: conn=[%p]\n", __func__, srv_conn);
+	WARN_ON(!spin_is_locked(&srv_conn->msg_qlock));
+	BUG_ON(list_empty(&srv_conn->msg_queue));
+
+	__tfw_http_req_fwd_stalled(srv_conn, &equeue);
+	spin_unlock(&srv_conn->msg_qlock);
+
+	if (!list_empty(&equeue))
+		tfw_http_req_zap_error(&equeue);
+}
+
+/*
+ * Forward the request @req to server connection @srv_conn.
+ *
+ * The request is added to the server connection (forwarding) queue.
+ * If forwarding is on hold at this moment, then the request will be
+ * forwarded later. Otherwise, if the queue is drained, then forward
+ * the request to the server immediately. If the queue is not drained,
+ * then forward all stalled requests to the server.
+ *
+ * Forwarding to a server is considered to be on hold after
+ * a non-idempotent request is forwarded to the server. The hold
+ * is removed when the holding non-idempotent request is followed
+ * by another request from the same client, which enables pipelining.
+ */
+static void
+tfw_http_req_fwd(TfwConnection *srv_conn, TfwHttpReq *req)
+{
+	bool drained;
+
+	TFW_DBG2("%s: srv_conn=[%p], req=[%p]\n", __func__, srv_conn, req);
+	BUG_ON(!(TFW_CONN_TYPE(srv_conn) & Conn_Srv));
+
+	spin_lock(&srv_conn->msg_qlock);
+	drained = tfw_http_conn_drained(srv_conn);
+	list_add_tail(&req->msg.fwd_list, &srv_conn->msg_queue);
+	atomic_inc(&srv_conn->qsize);
+	if (tfw_http_req_is_nonidempotent(req))
+		__tfw_http_req_nonidemp_enlist(srv_conn, req);
+	if (tfw_http_conn_on_hold(srv_conn)) {
+		spin_unlock(&srv_conn->msg_qlock);
+		TFW_DBG2("%s: Server connection is on hold: conn=[%p]\n",
+			 __func__, srv_conn);
+		return;
+	}
+	if (!drained) {
+		TFW_DBG2("%s: Server connection is not drained: conn=[%p]\n",
+			 __func__, srv_conn);
+		tfw_http_req_fwd_stalled(srv_conn);
+		/* The queue is unlocked inside the function. */
+		return;
+	}
+	if (tfw_connection_send(srv_conn, (TfwMsg *)req)) {
+		tfw_http_req_nonidemp_delist(srv_conn, req);
+		list_del_init(&req->msg.fwd_list);
+		atomic_dec(&srv_conn->qsize);
+		spin_unlock(&srv_conn->msg_qlock);
+		TFW_DBG2("%s: Forwarding error: conn=[%p] req=[%p]\n",
+			 __func__, srv_conn, req);
+		tfw_http_send_500(req);
+		TFW_INC_STAT_BH(clnt.msgs_otherr);
+		return;
+	}
```

Comment:
Yes, that's correct, and I did see that. The logic here was to have a special case for just one request which is not a rare case. That would avoid the extra logic related to sending a bunch of requests, like loop set up, etc, etc. Do you think this special case is unnecessary?